docs/discussions/server-side-graphql.md

```diff
@@ -0,0 +1,123 @@
+<!--[meta]
+section: discussions
+title: Server-side GraphQL Operations
+[meta]-->
+
+# Server-side GraphQL Operations
+
+Keystone provides a GraphQL API which we generally think of as being an API that is accessed over HTTP from a client talking to a server.
+There are many uses cases however where we want to execute GraphQL operations directly on the server.
+These can be grouped into two main categories:
+
+As part of a resolver, e.g.
+
+- Hooks
+- Access control
+- Virtual field resolver
+- Custom field type resolvers
+- Custom query/mutation resolvers
+
+As part of some other process, e.g.
+
+- Migrations
+- Data seeding
+- Responding to a non-GraphQL HTTP request
+- Testing
+- Anything else a developer could possibly imagine
+
+Keystone provides an API which give you complete control over how you execute server-side GraphQL operations in each of these situations.
+
+<!-- At the moment it is not obvious how a developer should go about executing a query in each of these situations. Ideally, we would have a single answer which could cover all of these bases. I would like the question "how do I execute a query in _situation X_ " will have a simple, obvious, and consistent answer.
+
+In this discussion I'd like to put forward a proposal for what that answer might be, and outline the changes required to implement such a solution. -->
+
+## Background
+
+There are four main things that go into executing a server-side GraphQL operation:
+
+- A GraphQL schema
+- The query string
+- Variable values
+- A `context` object
+
+When executing queries over HTTP we only have control over the queries and variables.
+On the server side we can also control the schema and the `context`.
+
+In most Keystone systems there will only be a single schema, but it is possible to configure multiple schemas (although this is not well documented).
+
+The [`context` object](https://www.apollographql.com/docs/apollo-server/data/resolvers/#the-context-argument) is a general purpose object which is made available to all resolvers. When performing HTTP queries this object is built by Apollo server based on the `request` object. When directly performing server-side operations we need to be able to create our own `context` object.
+
+## Within Resolvers
+
+While Keystone generates the basic CRUD resolvers for you, there are lots of ways you can write custom code which is executed during the resolver phase of a query.
+These include:
+
+- Hooks
+- Access control
+- Virtual field resolver
+- Custom field type resolvers
+- Custom query/mutation resolvers
+
+As a developer you might want to execute some GraphQL operations as part of your custom code.
```

Comment:
You'd get a little confused at this point if you don't know how a context object is provided to a resolved function I think we need to describe or show this before this point.

Reply:
Hmm, I wonder if we should just link out to the apollo docs at that point. If you're writing a custom resolver and you don't know about `context` then... 🤷 Maybe an simple example of using it in a hook and within access control, since those aren't full blown resolvers, but they have happen to have access to the context object, since they're called as part of the parent resolver.

Reply:
Or maybe we flip the order of this paragraph with the following one, where `context` objects are explained. I had something like that originally but then flipped it to this, but maybe that made it worse. 🤔

Reply:
This is also available in our hooks and stuff too right? It's less clear at that point that this is an Apollo context object.

Reply:
I've approved. Just this one general comment that dosen't need to block it.